src/czar/CzarRegistry.h

```diff
@@ -0,0 +1,127 @@
+/// in UserQueryFactory::newUserQuery and will be treated in similar manner as not being
+/// able to contact a worker.
+///
+/// There really shouldn't be communications problems, but there are, the best course of
+/// action would probably be to destroy the first instance of this and create a new one.
+///
+class CzarRegistry {
+public:
+    using Ptr = std::shared_ptr<CzarRegistry>;
+
+    /// Return a pointer to a new CzarRegistry object.
+    static Ptr create(std::shared_ptr<cconfig::CzarConfig> const& czarConfig) {
+        return Ptr(new CzarRegistry(czarConfig));
+    }
+
+    ~CzarRegistry();
+
+    struct WorkerContactInfo {
+        WorkerContactInfo(std::string const& wId_, std::string const& wHost_,
+                          std::string const& wManagementHost_, int wPort_, TIMEPOINT updateTime_)
+                : wId(wId_),
+                  wHost(wHost_),
+                  wManagementHost(wManagementHost_),
+                  wPort(wPort_),
+                  updateTime(updateTime_) {}
+        std::string const wId;              ///< key
+        std::string const wHost;            ///< "host-addr" entry.
+        std::string const wManagementHost;  ///< "management-host-name" entry.
+        int const wPort;                    ///< "management-port" entry.
+        TIMEPOINT const updateTime;         ///< "update-time-ms" entry.
+
+        /// Return true if all members, aside from updateTime, are equal.
+        bool same(WorkerContactInfo const& other) const {
+            return (wId == other.wId && wHost == other.wHost && wManagementHost == other.wManagementHost &&
+                    wPort == other.wPort);
+        }
+    };
+
+    using WorkerContactMap = std::unordered_map<std::string, WorkerContactInfo>;
+    using WorkerContactMapPtr = std::shared_ptr<WorkerContactMap>;
+
+private:
+    CzarRegistry() = delete;
+    CzarRegistry(std::shared_ptr<cconfig::CzarConfig> const& czarConfig);
+
+    /// This function will keep periodically updating Czar's info in the Replication System's Registry
+    /// until _loop is set to false.
+    /// Communications problems are logged but ignored. This should probably change.
+    void _registryUpdateLoop();
+
+    /// This function collects worker contact information from the Replication System's Registry
+    /// until _loop is set to false.
+    /// Communications problems are logged but ignored. This should probably change.
+    void _registryWorkerInfoLoop();
+
+    /// Build a new WorkerContactMap from the json `response`
+    WorkerContactMapPtr _buildMapFromJson(nlohmann::json const& response);
+
+    /// Return true if maps are the same size and all of the elements are the same().
+    bool _compareMap(WorkerContactMap const& other) const;
+
+    std::shared_ptr<cconfig::CzarConfig> const _czarConfig;  ///< Pointer to the CzarConfig.
+
+    std::atomic<bool> _loop{true};    ///< Threads will continue to run until this is set false.
```

Comment:
Is this really needed? `CzarRegistry` gets created by `Czar`, and they both would be gone when `Czar` is gone. The only reason for Czar to be destroyed is when a process is about to die. Who would care about a couple of 2 extra threads when there could be hundreds of other query-processing threads? The threads won't prevent the process from dying, and there is nothing useful these threads could do before the end. The controlled termination of threads only makes sense if/when there is a plan for cleaning up resources or recreating `CzarRegistry` while `Czar` is still alive.

Nothing is bad about the implementation. Just curious.

Reply:
I don't like using detached threads as I've I have had real trouble with race condiftions in unit tests. The detached thread accesses something and you get segfault depending on what stopped first. Using _loop and join() results in repeatable termination conditions.

Reply:
If you will design your class as:
```
class CzarRegistry : public std::enable_shared_from_this<Server> {
```
And use the shared pointers instead of the low-level `this` then you won't have any segfaults or race conditions.